app/src/test/java/org/mozilla/fenix/wifi/WifiConnectionMonitorTest.kt

```diff
@@ -0,0 +1,192 @@
+
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener({})
+
+        assertFalse(wifiConnectionMonitor.callbacks.isEmpty())
+    }
+
+    @Test
+    fun `addOnWifiConnectedChangedListener AND lastKnownState different of NULL - SHOULD register callback and call it`() {
+        var wasNotified: Boolean? = null
+        wifiConnectionMonitor.lastKnownStateWasAvailable = false
+
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener {
+            wasNotified = it
+        }
+
+        assertFalse(wasNotified!!)
+        assertFalse(wifiConnectionMonitor.callbacks.isEmpty())
+    }
+
+    @Test
+    fun `removeOnWifiConnectedChangedListener - SHOULD remove callback`() {
+
+        assertTrue(wifiConnectionMonitor.callbacks.isEmpty())
+
+        val callback: (Boolean) -> Unit = {}
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener(callback)
+
+        assertFalse(wifiConnectionMonitor.callbacks.isEmpty())
+
+        wifiConnectionMonitor.removeOnWifiConnectedChangedListener(callback)
+
+        assertTrue(wifiConnectionMonitor.callbacks.isEmpty())
+    }
+
+    @Test
+    fun `frameworkListener - SHOULD notify when connection lost`() {
+        var wasNotified: Boolean? = null
+
+        val callback: (Boolean) -> Unit = { wasNotified = it }
+
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener(callback)
+        wifiConnectionMonitor.frameworkListener.onLost(mock())
+
+        assertFalse(wasNotified!!)
+    }
+
+    @Test
+    fun `frameworkListener - SHOULD notify when connection available`() {
+        var wasNotified: Boolean? = null
+
+        val callback: (Boolean) -> Unit = { wasNotified = it }
+
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener(callback)
+        wifiConnectionMonitor.frameworkListener.onAvailable(mock())
+
+        assertTrue(wasNotified!!)
+    }
+
+    @Test
+    fun `notifyListeners - SHOULD notify listeners`() {
+        var wasNotified1: Boolean? = null
+        var wasNotified2: Boolean? = null
+
+        val callback1: (Boolean) -> Unit = { wasNotified1 = it }
+        val callback2: (Boolean) -> Unit = { wasNotified2 = it }
+
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener(callback1)
+        wifiConnectionMonitor.addOnWifiConnectedChangedListener(callback2)
+        wifiConnectionMonitor.notifyListeners(true)
+
+        assertTrue(wasNotified1!!)
+        assertTrue(wasNotified2!!)
+    }
+
+    @Test
+    fun `notifyListeners - SHOULD NOT throw a ConcurrentModificationException`() {
+        repeat(100) {
+
+            // Adding to callbacks.
+            wifiConnectionMonitor.addOnWifiConnectedChangedListener {
+                // Altering callbacks while looping.
+                if (wifiConnectionMonitor.callbacks.isNotEmpty())
+                    wifiConnectionMonitor.callbacks.removeFirst()
+            }
+
+            // Looping over callbacks.
+            wifiConnectionMonitor.notifyListeners(true)
+        }
+    }
```

Comment:
@csadilek this should be covering the missing test for this pr https://github.com/mozilla-mobile/fenix/pull/16372